### PR TITLE
Added Kline data API for Binance

### LINKED ISSUE #642: Adding support for Binance Crypto Currency Exchange
- [Binance Exchange](https://www.binance.com/en) is a exchange for crypto currencies.
- It provides a API for users to retrieve data
- We can add support for this which would increase the usability of this module

### PR DESCRIPTION
**Pull request by rohitjain00**
- [ ] closes #642
- [x] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] passes `black --check pandas_datareader`
- [ ] added entry to docs/source/whatsnew/vLATEST.txt
- [ ] added documentation for new reader

### CODE REVIEW

File: pandas_datareader/binance/__init__.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        if self.start is not None:
+            p["startTime"] = self.convert_time_to_miliseconds(self.start)
+        if self.end is not None:
+            p["endTime"] = self.convert_time_to_miliseconds(self.end)
+
+        return p
+
+    def read(self):
+        data = requests.get(url=self.url, params=self.params)
```
Comment (addisonlynch) [inline pandas_datareader/binance/__init__.py]: ``BaseReader`` can handle this. Try the following:

```python
def read(self):
    data = super(BinanceReader, self).read()
    return self.clean_data(data)
```

You can also remove the ``requests`` import

File: pandas_datareader/binance/__init__.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        )
+        self._interval = interval
+        self._limit = limit
+
+    @property
+    def url(self):
+        """API URL"""
+        return BINANCE_BASE_URL + "/api/v1/klines"
+
+    def clean_data(self, data):
```
Comment (addisonlynch) [inline pandas_datareader/binance/__init__.py]: Add leading underscore to this method as it's only used internally (``_clean_data``)

File: pandas_datareader/binance/__init__.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                "ignore",
+            ],
+        )
+        dataFrame = dataFrame.drop(columns="ignore")
+        return dataFrame
+
+    def get_interval(self):
+        return self._interval
+
+    def convert_time_to_miliseconds(self, dt):
```
Comment (addisonlynch) [inline pandas_datareader/binance/__init__.py]: This is not python 2 compatible. Use ``time`` instead:

```python
int(time.mktime(dt.timetuple()) * 1000)
```

Comment (addisonlynch) [reply pandas_datareader/binance/__init__.py]: Also same comment re: leading underscore

File: pandas_datareader/tests/binance/test_binance.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            self.columns_should_be_present
+        ), "Number of columns in the returned data is higher than 11"
+        assert (
+            self.binance_reader.shape[0] <= self.limit
+        ), "Number of rows in data returned is higher than the expected number"
+
+    def test_object_columns(self):
+        for col in self.binance_reader.columns:
+            assert col in self.columns_should_be_present, (
+                "Column" + col + "should not be present in the response"
```
Comment (addisonlynch) [inline pandas_datareader/tests/binance/test_binance.py]: These are OK tests for testing the formatting of the data returned...is it possible to add some tests of values?

Comment (rohitjain00) [reply pandas_datareader/tests/binance/test_binance.py]: Like specified in the Issue...I can check for values on a particular date or for a span of time.

Comment (addisonlynch) [reply pandas_datareader/tests/binance/test_binance.py]: Particular dates are fine.

File: pandas_datareader/binance/__init__.py (hunk 1 above)
Comment (rohitjain00) [reply pandas_datareader/binance/__init__.py]: The response for the API call is : 
Screenshot from 2019-09-23 18-53-01
So I cannot specify a format to read this data. Hence a `NotImplementedError` is occurred.

**Comment by bashtage** [comment]
@rohitjain00 I added a task that is only relevant for new readers -- docs.

File: pandas_datareader/tests/binance/test_binance.py (hunk 4 above)
Comment (rohitjain00) [reply pandas_datareader/tests/binance/test_binance.py]: I've added a test to check the data of "BNBBTC" on 17th May 2018.

File: pandas_datareader/binance/__init__.py (hunk 2 above)
Comment (rohitjain00) [reply pandas_datareader/binance/__init__.py]: I have successfully changed this to private and tested the code.

File: pandas_datareader/binance/__init__.py (hunk 3 above)
Comment (rohitjain00) [reply pandas_datareader/binance/__init__.py]: This has been updated and checked.

**Comment by rohitjain00** [comment]
> Thanks. At first glance, a couple things:
> 
> * Can you add some documentation for this reader? To both whatsnew (may need to create a file for 0.9.0) and `remote-data.rst`
> * Use [isort](https://github.com/timothycrosley/isort) to reorder/reformat imports

I can add the documentation once the PR gets merge with the project.

- In [`remote_data.rst`](https://github.com/pydata/pandas-datareader/blob/master/docs/source/remote_data.rst) I need to add a sample code of calling the method. Write what are the possible values of a parameter
    - For example in this `intervals` parameter takes only specified input ([Specified here](https://github.com/rohitjain00/pandas-datareader/blob/bdde4c5585a5282d9fda0892ff469fe36fef6f0d/pandas_datareader/binance/__init__.py#L16))

- In [whatsnew v0.8.0](https://github.com/pydata/pandas-datareader/blob/master/docs/source/whatsnew/v0.8.0.txt#L16) I can add just that we have a new Binance support for extracting kline data of different Crypto currrencies

- As for [isort](https://github.com/timothycrosley/isort) I have reordered the imports. I feel that we can create a script file which would reorder the imports on all files using `isort` and format the code using `black`

**Comment by gliptak** [comment]
@rohitjain00 please update https://github.com/pydata/pandas-datareader/blob/master/docs/source/remote_data.rst as part of this PR

**Comment by rohitjain00** [comment]
Should I update in [Whatsnew](https://github.com/pydata/pandas-datareader/blob/master/docs/source/whatsnew/v0.8.0.txt#L16) file?

v 0.8.1 is just released so I think this feature is supposed to be added in v 0.8.2.

**Comment by addisonlynch** [comment]
@rohitjain00 it will be 0.9.0.

**Comment by rohitjain00** [comment]
@gliptak Should I add a v0.9.0.txt file to whatsnew folder?

**Comment by bashtage** [comment]
Yes.  Please copy the 0.8 one and remove the contents of each section to use as a template.

File: docs/source/remote_data.rst [hunk 5]
```diff
@@ -682,3 +682,78 @@ The Moscow Exchange (MOEX) provides historical data.
...
+
+Binance Data
+=========
+The Binance API provides the recent Kline/Candlestick data for cryptocurrency.
+
+
+
+.. code:: ipython3
+
+    import pandas_datareader as pdr
```
Comment (bashtage) [inline docs/source/remote_data.rst]: Could you use one block here,
```
.. code:: ipython

   line1
   line2
   line3
```

File: docs/source/remote_data.rst [hunk 6]
```diff
@@ -682,3 +682,78 @@ The Moscow Exchange (MOEX) provides historical data.
...
+
+.. code:: ipython3
+
+    data_binance
+
+
+
+
+.. raw:: html
+
```
Comment (bashtage) [inline docs/source/remote_data.rst]: Please convert to RST table.  You can do this using pandoc online.

**Comment by rohitjain00** [comment]
Is there anything missing for this PR?

**Comment by gliptak** [comment]
@bashtage (maybe beside squashing) are there additional updates you would need to see here?

**Comment by rohitjain00** [comment]
@bashtage Could you review this?

**Comment by rohitjain00** [comment]
@bashtage is this commit under consideration.?

**Comment by bashtage** [comment]
Yes, but I'm quite constrained for the next 4 weeks.

**Comment by rohitjain00** [comment]
@bashtage could you review this?

**Comment by alik604** [comment]
@rohitjain00 I'm sorry to read about your experience 😥

**Comment by jlpetracca** [comment]
@rohitjain00 is this working? @bashtage

**Pull request closed without merging**